lib/Sema/CSDiagnostics.cpp

```diff
@@ -1477,3 +1477,223 @@ bool MissingMemberFailure::diagnoseAsError() {
   corrections.noteAllCandidates();
   return true;
 }
+
+bool AllowTypeOrInstanceMemberFailure::diagnoseAsError() {
+  auto loc = getAnchor()->getLoc();
+  auto &tc = getTypeChecker();
+  auto &cs = getConstraintSystem();
+  
+  Expr *expr = getParentExpr();
+  SourceRange baseRange = expr ? expr->getSourceRange() : SourceRange();
+  auto member = getAnchor()->getReferencedDecl().getDecl();
+  
+  // If the base is an implicit self type reference, and we're in a
+  // an initializer, then the user wrote something like:
+  //
+  //   class Foo { let x = 1, y = x }
+  //
+  // which runs in type context, not instance context, or
+  //
+  //   class Bar {
+  //     let otherwise = 1              // instance member
+  //     var x: Int
+  //     func init(x: Int =otherwise) { // default parameter
+  //       self.x = x
+  //     }
+  //   }
+  //
+  // in which an instance member is used as a default value for a
+  // parameter.
+  //
+  // Produce a tailored diagnostic for these cases since this
+  // comes up and is otherwise non-obvious what is going on.
+  
+  if (BaseType->is<MetatypeType>()) {
+    auto instanceTy = BaseType->getRValueType();
+    
+    if (auto *AMT = instanceTy->getAs<AnyMetatypeType>()) {
+      instanceTy = AMT->getInstanceType();
+    }
+    
+    if (expr && expr->isImplicit() && cs.DC->getContextKind() == DeclContextKind::Initializer) {
+      auto *TypeDC = cs.DC->getParent();
+      bool propertyInitializer = true;
+      // If the parent context is not a type context, we expect it
+      // to be a defaulted parameter in a function declaration.
+      if (!TypeDC->isTypeContext()) {
+        assert(TypeDC->getContextKind() ==
+               DeclContextKind::AbstractFunctionDecl &&
+               "Expected function decl context for initializer!");
+        TypeDC = TypeDC->getParent();
+        propertyInitializer = false;
+      }
+      
+      assert(TypeDC->isTypeContext() && "Expected type decl context!");
+      
+      if (TypeDC->getSelfNominalTypeDecl() == instanceTy->getAnyNominal()) {
+        if (propertyInitializer) {
+          tc.diagnose(loc, diag::instance_member_in_initializer,
```

Comment:
Instead of using `TypeChecker::diagnose` directly please switch to `emitDiagnostic` which is already provided by `FailureDiagnostic`